### ISSUE TITLE
docker-compose + persistent data volume (2.8.2+)

### ISSUE THREAD

**Issue by jackross** [arangodb/arangodb-docker#19]
The following `docker-compose.yml` works just fine with **2.8.1**:

``` yaml
# docker-compose.yml
app:
  image: arangodb:2.8.1
  volumes:
    - ./apps:/var/lib/arangodb-apps/_db
  volumes_from:
    - arangodb_data
  ports:
    - 8529:8529
  environment:
    - ARANGO_NO_AUTH=1

arangodb_data:
  image: tianon/true
  volumes:
    - /var/lib/arangodb
```

``` sh
$ docker-compose up
Starting arangodb_arangodb_data_1
Starting arangodb_app_1
Attaching to arangodb_arangodb_data_1, arangodb_app_1
app_1           | 2016-02-22T16:02:25Z [1] INFO ArangoDB 2.8.1 [linux] 64bit, using tcmalloc, ICU 54.1, V8 4.3.61, OpenSSL 1.0.1k 8 Jan 2015
app_1           | 2016-02-22T16:02:25Z [1] INFO using default language 'en'
app_1           | 2016-02-22T16:02:25Z [1] INFO loaded database '_system' from '/var/lib/arangodb/databases/database-127711'
app_1           | 2016-02-22T16:02:25Z [1] INFO running WAL recovery (1 logfiles)
app_1           | 2016-02-22T16:02:25Z [1] INFO replaying WAL logfile '/var/lib/arangodb/journals/logfile-19994569107.db' (1 of 1)
app_1           | 2016-02-22T16:02:25Z [1] INFO WAL recovery finished successfully
app_1           | 2016-02-22T16:02:25Z [1] INFO using endpoint 'tcp://0.0.0.0:8529' for non-encrypted requests
app_1           | 2016-02-22T16:02:25Z [1] INFO using default API compatibility: 20800
app_1           | 2016-02-22T16:02:25Z [1] INFO JavaScript using startup '/usr/share/arangodb/js', application '/var/lib/arangodb-apps'
arangodb_arangodb_data_1 exited with code 0
app_1           | 2016-02-22T16:02:27Z [1] INFO In database '_system': Database is up-to-date (20801/prod/standalone/existing)
app_1           | 2016-02-22T16:02:29Z [1] INFO Please note that a new bugfix version '2.8.2' is available
app_1           | 2016-02-22T16:02:29Z [1] INFO file-descriptors (nofiles) hard limit is 1048576, soft limit is 1048576
app_1           | 2016-02-22T16:02:29Z [1] INFO Authentication is turned off
app_1           | 2016-02-22T16:02:29Z [1] INFO ArangoDB (version 2.8.1 [linux]) is ready for business. Have fun!
```

However, the same `docker-compose.yml` file (changed to point to **2.8.2**) does not work:

``` yaml
# docker-compose.yml
app:
  image: arangodb:2.8.2
  volumes:
    - ./apps:/var/lib/arangodb-apps/_db
  volumes_from:
    - arangodb_data
  ports:
    - 8529:8529
  environment:
    - ARANGO_NO_AUTH=1

arangodb_data:
  image: tianon/true
  volumes:
    - /var/lib/arangodb
```

``` sh
$ docker-compose up
Creating arangodb_arangodb_data_1
Creating arangodb_app_1
Attaching to arangodb_arangodb_data_1, arangodb_app_1
arangodb_arangodb_data_1 exited with code 0
app_1           | chown: changing ownership of '/var/lib/arangodb-apps/_db': Operation not permitted
arangodb_app_1 exited with code 1
```

I am using the latest `docker`, `docker-compose` and `docker-machine` on OSX Yosemite with VMWare Fusion.

``` sh
$ docker --version
Docker version 1.10.1, build 9e83765
$ docker-compose --version
docker-compose version 1.6.0, build unknown
$ docker-machine --version
docker-machine version 0.6.0, build e27fb87
```

**Comment by m0ppers** [comment]
Fail! We changed our docker setup for 2.8.2 and are using a dedicated user "arangodb" since 2.8.2. Upon startup we are ensuring proper access (rw) to our data and apps directories. This is currently done using chown. This works properly when using Linux.

When using a Mac there is however one more indirection as the mac has to first share the `apps` directory with the docker machine. A chown from inside the docker container will then fail even though the setup runs as root because the chown is not allowed to pass from the docker machine to your local machine (via shared folder functionality).

Long story short: We will change the chown to a simple "are we allowed to write" check. We are currently creating 2.8.3 and will fix it there :)

2.8.3 should be available the next days :)

**Comment by jackross** [comment]
Who knew the most critical piece of information provided was the very last piece (OS X).  It definitely pays to provide as much info as possible.  Thanks for following up!

**Comment by jackross** [comment]
@m0ppers  Finally got a chance to play with **2.8.3** today, and I'm still encountering issues, although the issues did change slightly as a result of your changes to the 2.8.3 release.  To recap, using the following `docker-compose.yml`:

``` yml
app:
  image: arangodb:2.8.3
  volumes_from:
    - arangodb_data
  ports:
    - 8529:8529
  environment:
    - ARANGO_NO_AUTH=1

arangodb_data:
  image: tianon/true
  volumes:
    - /var/lib/arangodb
```

and then running:

``` sh
docker-compose up
```

I get the following:

``` sh
Creating cdparangodb_arangodb_data_1
Creating cdparangodb_app_1
Attaching to cdparangodb_arangodb_data_1, cdparangodb_app_1
app_1           | touch: cannot touch '/var/lib/arangodb/_rwcheck_80b3df00987e': Permission denied
app_1           | We seem to not have proper rw access to /var/lib/arangodb. Please make sure that every mounted volume has full rw access for user arangodb (uid=999(arangodb) gid=999(arangodb) groups=999(arangodb))
cdparangodb_arangodb_data_1 exited with code 0
cdparangodb_app_1 exited with code 55
```

which, looking through the entrypoint script, seems to make sense...

I should add that I get this on Mac OS X, **_as well as on an Ubuntu 15.10 machine directly**_.  Not saying that there is a problem, but could you direct me as to how I could set up permissions correctly so as to use a mounted data only volume?

**Comment by m0ppers** [comment]
I forgot to mention something important. Sorry.

As we are now leaving out any chowns you are responsible to setup proper permissions. In your first attempt you were mounting the apps directory from your local mac. So to make that work you just need to make sure that uid,gid 999 is allowed to write there on your local mac (chmod 777 for the lazy :D).

The data volume however is coming from the tianon/true image. This image doesn't know anything about the arangodb user needing permission on the data directory so the volume will be created as root and PENG arangodb can't write to it.

So to make that work the following change to the data image should work:

```
arangodb_data:
  image: arangodb:2.8.3
```

The image already contains the volume /var/lib/arangodb so that can be omitted. What is furthermore happening when using the arangodb image: it will come with proper uid/gid settings for that directory (https://github.com/arangodb/arangodb-docker/blob/official/Dockerfile.templ#L38)

Hope that works for you. Sorry this is a great usability madness. The only way to fix that which comes to my mind is going back to run arangodb as root bu that is certainly not best practice. If somebody comes up with a more useable solution we would be happy to incorporate it :(

**Comment by jackross** [comment]
Thanks @m0ppers.  That works great, all around.  Appreciate the help!

**Comment by sartaj** [comment]
Hi @m0ppers thanks for the heads up.

Until a better use case is resolved, I'd highly recommend adding instructions on what to do on Docker Hub.

**Comment by dothebart** [comment]
https://hub.docker.com/_/arangodb/ mentions this under "persistant data"
